Add docs page on high swap usage (#20027)
supabase/supabase: apps/docs/pages/guides/platform/exhaust-swap.mdx

### When is high Swap concerning?

High Swap is concerning if your instance is using all of the available RAM (i.e. consistently using more than 75%). 

High Swap usage could cause noticeable performance impacts to your instance, typically this could manifest as:

- **Slower query responses.** 
- **Degraded performance due to swapping regularly between RAM and disk.**
- **Higher Disk I/O due to swapping regularly.**

## Monitor your Swap

You can check your Swap usage directly on the Supabase Platform. For this go to database health in the reports section or [click here](https://supabase.com/dashboard/project/_/reports/database) and select your project.

It is also possible to monitor your resources and set up alerts using Prometheus/Grafana. You can find a guide for this [here](https://supabase.com/docs/guides/platform/metrics).

An [example repository](https://github.com/supabase/supabase-grafana) to ingest metrics and visualize them with Grafana is provided in the linked guide, where we maintain a [list of the exported metrics](https://github.com/supabase/supabase-grafana/blob/main/docs/metrics.md).

Some useful metrics to monitor are:

> - `node_memory_SwapFree_bytes` - Shows the total amount of Swap available in bytes.

Reviewer: Mention the ones for RAM utilization as well? And page fault rate? Page fault rate might not be included in older projects thougth

- `node_disk_io_time_seconds_total` and `node_disk_io_now` - Shows the amount of time spent on disk I/O, an increase in this **could** mean excessive swapping.

## Common reasons for high Swap usage

> Everything you do with your Supabase project requires compute. Hence, there can be many reasons for high Swap usage. Here are some common ones:

Reviewer: Add mention and link to page about reducing the number of connections in use?

Author: I think that is linked to in line 65 isn't it? (Assuming you mean this guide: https://supabase.com/docs/guides/platform/performance#optimizing-the-number-of-connections)

Reviewer: Yeah it's on the same page, I was wondering if its worth calling it out explicitly as one of the causes for high memory utilization. Fine either way

import Layout from '~/layouts/DefaultGuideLayout'

export const meta = {
  id: 'exhaust-swap',
  title: 'High Swap Usage',
  description:
    'Learn what high Swap usage could mean for your Supabase instance and what could have caused it.',
}

> Learn what high Swap usage could mean for your Supabase instance and what could have caused it.

Reviewer 2: ```suggestion
Learn what high Swap usage means, what can cause it, and how to solve it.
```

## What is Swap for?

<Admonition type="tip">

> High Swap is generally not a concern if you find other resources on your instance are not constrained.

Reviewer 2: ```suggestion
High Swap is usually not a problem unless other resources (such as RAM) are constrained.
```

</Admonition>

> Every Supabase project runs in its dedicated virtual machine. Your instance will have a different set of hardware provisioned depending on your [compute add-on](https://supabase.com/docs/guides/platform/compute-add-ons). Your hardware may not be suitable for the intended workload and may experience high Swap usage.

Reviewer 3: ```suggestion
Every Supabase project runs in its own dedicated virtual machine. Your instance will have a different set of hardware provisioned depending on your [compute add-on](https://supabase.com/docs/guides/platform/compute-add-ons). Your hardware may not be suitable for the intended workload and may experience high Swap usage.
```

> Every Supabase project runs in its own dedicated virtual machine. Your instance will have a different set of hardware provisioned depending on your [compute add-on](https://supabase.com/docs/guides/platform/compute-add-ons). Your hardware may not be suitable for the intended workload and may experience high Swap usage.

Reviewer 2: ```suggestion
Every Supabase project runs on its own dedicated virtual machine. The machine's underlying specs and hardware depend on your [compute add-on](https://supabase.com/docs/guides/platform/compute-add-ons). If your hardware isn't suitable for your workload, you might experience high Swap usage.
```

Swap is a portion of your instance's disk that is reserved for the operating system to use when the available RAM has been utilized. As it uses the disk, Swap is slower to access and is generally used as a last resort.

> Swap can be used even if your instance has plenty of RAM. If this is the case, do not worry.
> 
> Your instance might try to "preemptively swap" by swapping background processes to make space for your traffic in RAM.

Reviewer 2: ```suggestion
Swap can be used even if your instance has plenty of RAM. If this is the case, do not worry. Your instance might try to "preemptively swap" by swapping background processes to make space for your traffic in RAM.
```

### When is high Swap concerning?

High Swap is concerning if your instance is using all of the available RAM (i.e. consistently using more than 75%). 

> High Swap usage could cause noticeable performance impacts to your instance, typically this could manifest as:

Reviewer 2: ```suggestion
High Swap usage can affect your database performance. For example, you might see:
```

- **Slower query responses.** 
- **Degraded performance due to swapping regularly between RAM and disk.**
- **Higher Disk I/O due to swapping regularly.**

## Monitor your Swap

> You can check your Swap usage directly on the Supabase Platform. For this go to database health in the reports section or [click here](https://supabase.com/dashboard/project/_/reports/database) and select your project.

Reviewer 2: ```suggestion
You can check your Swap usage directly on the Supabase Platform. Navigate to the [**Database** page](https://supabase.com/dashboard/project/_/reports/database) of the **Reports** section.
```

> It is also possible to monitor your resources and set up alerts using Prometheus/Grafana. You can find a guide for this [here](https://supabase.com/docs/guides/platform/metrics).

Reviewer 2: ```suggestion
You can also monitor your resources and set up alerts using Prometheus/Grafana. See the [metrics guide](https://supabase.com/docs/guides/platform/metrics) for more information.
```

An [example repository](https://github.com/supabase/supabase-grafana) to ingest metrics and visualize them with Grafana is provided in the linked guide, where we maintain a [list of the exported metrics](https://github.com/supabase/supabase-grafana/blob/main/docs/metrics.md).

Some useful metrics to monitor are (this is not an exhaustive list):

> - `node_memory_SwapFree_bytes` - Shows the total amount of Swap available in bytes.

Reviewer 2: ```suggestion
- `node_memory_SwapFree_bytes` - The total amount of Swap available in bytes.
```

> - `node_disk_io_time_seconds_total` and `node_disk_io_now` - Shows the amount of time spent on disk I/O, an increase in this **could** mean excessive swapping.

Reviewer 2: ```suggestion
- `node_disk_io_time_seconds_total` and `node_disk_io_now` - The amount of time spent on disk I/O. An increase might be an indirect sign of excessive swapping, but not always.
```

> - `node_memory_MemTotal_bytes` and `node_memory_MemFree_bytes` - Shows the total RAM and amount of available RAM on your instance

Reviewer 2: ```suggestion
- `node_memory_MemTotal_bytes` and `node_memory_MemFree_bytes` - The total RAM and available RAM.
```

> - `node_vmstat_pswpin` and `node_vmstat_pswpout` - Shows the number of pages that have been swapped in or out (monitoring this for spikes means that your instance is swapping)

Reviewer 2: ```suggestion
- `node_vmstat_pswpin` and `node_vmstat_pswpout` - The number of pages that have been swapped in or out (monitoring this for spikes means that your instance is swapping).
```

## Common reasons for high Swap usage

Everything you do with your Supabase project requires compute. Hence, there can be many reasons for high Swap usage. Here are some common ones:

> - **Query performance:** You might have high read traffic or you may have queries that have to process a large amount of data on disk (this can happen even if you are returning a small amount of data).

Reviewer 2: ```suggestion
- **Query performance:** You might have high read traffic or queries that process a large amount of data on disk (this can happen even if you are returning a small amount of data).
```

> - **Missing indexes:** Your database might have to scan through a large amount of data to find the information it needs. Creating indexes helps your database find data faster. Learn more about indexes [here](https://supabase.com/docs/guides/database/postgres/indexes).

Reviewer 2: ```suggestion
- **Missing indexes:** Your database might have to scan through a large amount of data to find the information it needs. Creating indexes helps your database find data faster. See the [indexes guide](https://supabase.com/docs/guides/database/postgres/indexes) to learn more.
```

- **Unsuitable compute:** The instance size of your Supabase project might not be suitable for your application as you might have more traffic or run resource-intensive operations.

> - **Workload style:** The usage pattern of your Supabase project might be more read heavy, or involve large amounts of data, that can increase Swap usage.

Reviewer 2: ```suggestion
- **Workload style:** The usage pattern of your Supabase project might be more read heavy, or involve large amounts of data.
```

> - **Extensions:** You may be using extensions that enable you to get the full power from your Supabase database, this may come with higher resource usage if they are performing intensive operations on large datasets.

Reviewer 2: ```suggestion
- **Extensions:** You might be using extensions that perform intensive operations on large datasets. This increases resource usage.
```

## Solving high Swap usage

> If you find that your RAM and Swap usage are high, then you have three options:

Reviewer 2: ```suggestion
If you find that your RAM and Swap usage are high, you have three options:
```

> 1. **Optimize performance:** Get more out of your instance's resources by optimizing your usage. Have a look at our [performance tuning guide](https://supabase.com/docs/guides/platform/performance#examining-query-performance) and our [production readiness guide](https://supabase.com/docs/guides/platform/going-into-prod#performance).

Reviewer 2: ```suggestion
1. **Optimize performance:** Get more out of your instance's resources by optimizing your usage. See the [performance tuning guide](https://supabase.com/docs/guides/platform/performance#examining-query-performance) and our [production readiness guide](https://supabase.com/docs/guides/platform/going-into-prod#performance).
```

> 2. **Upgrade your compute:** You can get a Compute Add-on for your project. Follow [this link](https://supabase.com/dashboard/project/_/settings/addons?panel=computeInstance) and select your project to see your upgrade options.

Reviewer 2: ```suggestion
2. **Upgrade your compute:** You can get a Compute Add-on for your project. Navigate to [the **Addons** section in project settings](https://supabase.com/dashboard/project/_/settings/addons?panel=computeInstance) to see your upgrade options.
```

> 3. **Read Replicas:** You can spread the load on your Supabase project by creating a Read Replica, see [this link](https://supabase.com/docs/guides/platform/read-replicas) for more details

Reviewer 2: ```suggestion
3. **Read Replicas:** You can spread the load on your Supabase project by creating a Read Replica. See [the read replicas guide](https://supabase.com/docs/guides/platform/read-replicas) for more information.
```
